engine/api/src/main/scala/pl/touk/nussknacker/engine/api/definition/WithCustomValidation.scala

```diff
@@ -0,0 +1,12 @@
+package pl.touk.nussknacker.engine.api.definition
+
+import pl.touk.nussknacker.engine.api.expression.Expression
+import pl.touk.nussknacker.engine.api.typed.typing.TypingResult
+
+/**
+ * Extend if you want your service to perform additional validations after compilation
+ * e.g. parameters cross validation
+ */
+trait WithCustomValidation {
+  def validate(params: Map[String, (Expression, TypingResult)]): List[String]
```

Comment:
can we return here `ValidatedNel[List[String], Unit]` ?

Reply:
Good idea, but I've removed the whole trait